File: 2105.watering-plants-ii.py
#
# @lc app=leetcode id=2105 lang=python3
#
# [2105] Watering Plants II
#

# @lc code=start
from typing import List
class Solution:
    def minimumRefill(self, plants: List[int], capacityA: int, capacityB: int) -> int:

        #max(plants[i]) <= capacityA, capacityB <= 109
        left, right = 0, len(plants) - 1
        cnt = 0
        Alice, Bob = capacityA, capacityB
        while left < right:
            if Alice < plants[left]:
                Alice = capacityA
                cnt += 1
            if Bob < plants[right]:
                Bob = capacityB
                cnt += 1
            Alice  -= plants[left]
            Bob -= plants[right]
            left += 1
            right -= 1
        
        if left == right:
            if (Alice == Bob and Alice < plants[left]) or (Bob > Alice and Bob < plants[left]) or (Alice > Bob and Alice < plants[left]):
                cnt += 1
        return cnt

        
# @lc code=end

File: 2105.watering-plants-ii.java
/*
 * @lc app=leetcode id=2105 lang=java
 *
 * [2105] Watering Plants II
 */

// @lc code=start
class Solution {
    public int minimumRefill(int[] plants, int A, int B) {
        //最直观的双指针模拟
        int n = plants.length;
        int left = 0, right = n -1;
        int Alice = A, Bob = B;
        int ans = 0;
        while(left < right){
            if(Alice < plants[left]){
                ans++;
                Alice = A;
            }
            Alice -= plants[left];
            left++;
            if(Bob < plants[right]){
                ans++;
                Bob = B;
            }
            Bob -= plants[right];
            right--;
        }
        if(left == right && Math.max(Alice, Bob) < plants[left]) ans++;
        return ans;
    }
}
// @lc code=end

